allennlp/models/encoder_decoders/nlvr_semantic_parser.py

```diff
@@ -820,25 +1011,31 @@ def _compute_new_states(cls,  # type: ignore
                     continue
                 # If action is not in instance_agenda, mask_variable, and checklist_addition will be
                 # all 0s.
-                checklist_mask = instance_agenda == action  # (agenda_size,)
-                checklist_addition = checklist_mask.float() * action_prob  # (agenda_size,)
+                checklist_mask = (instance_agenda == action).float()  # (agenda_size,)
+                agenda_action_prob = torch.sum(checklist_mask * agenda_action_probs)  # (1,)
+                # TODO (pradeep): This is not a great way to bias the model towards choosing actions
+                # that fill the checklist. May be use the current checklist in the action logit
+                # computation itself.
+                action_prob = checklist_weight * agenda_action_prob + \
```

Comment:
Are you actually ending up with a valid probability distribution when you do this?  Also, my recommendation would be to put this logic in `_take_step`, and still have `log_probs` be the input to this function, so there's a more obvious distinction between where model computation happens and where state booking happens.

Reply:
Yes, it is a valid probability distribution. I agree with you that model computation should happen in `_take_step`. There are some changes in this method in the #952. I'll make the change you recommended there since that would be easier.